enhancements/cert-manager/istio-csr-controller.md

```diff
@@ -0,0 +1,879 @@
+spec:
+  isCA: true
+  duration: 21600h
+  secretName: istio-csr-ca
+  commonName: istio-csr-ca
+  issuerRef:
+    name: istio-csr-cluster-issuer
+    kind: ClusterIssuer
+    group: cert-manager.io
+---
+apiVersion: cert-manager.io/v1
+kind: Issuer
+metadata:
+  name: istio-csr-issuer
+  namespace: istio-system
+spec:
+  ca:
+    secretName: istio-csr-ca
+```
+
+3. Resources required for deploying istio-csr agent.
+   1. `istio-system` namespace, where all the istio-csr required resources will be created.
+      ```yaml
+      apiVersion: v1
+      kind: Namespace
+      metadata:
+        name: istio-system
+      ```
+
+   2. Service for creating istio-csr grpc server, for serving CertificateRequests endpoint.
+      ```yaml
+      apiVersion: v1
+      kind: Service
+      metadata:
+        name: cert-manager-istio-csr
+        namespace: istio-system
+        labels:
+          app: cert-manager-istio-csr
+      spec:
+        type: ClusterIP
+        ports:
+        - port: 443
+          targetPort: 6443
+          protocol: TCP
+          name: grpc
+        selector:
+          app: cert-manager-istio-csr
+      ```
+
+   3. Service for supporting metrics.
+      ```yaml
+      apiVersion: v1
+      kind: Service
+      metadata:
+        name: cert-manager-istio-csr-metrics
+        namespace: istio-system
+        labels:
+          app: cert-manager-istio-csr
+      spec:
+        type: ClusterIP
+        ports:
+        - port: 9402
+          targetPort: 9402
+          protocol: TCP
+          name: metrics
+        selector:
+          app: cert-manager-istio-csr
+      ```
+
+   4. ServiceMonitor resource to monitor and collect metrics of istio-csr.
+      ```yaml
+      apiVersion: monitoring.coreos.com/v1
+      kind: ServiceMonitor
+      metadata:
+        name: cert-manager-istio-csr
+        namespace: istio-system
+        labels:
+          app: cert-manager-istio-csr
+          prometheus: default
+      spec:
+        jobLabel: cert-manager-istio-csr
+        selector:
+          matchLabels:
+            app: cert-manager-istio-csr-metrics
+        namespaceSelector:
+          matchNames:
+            - cert-manager
+        endpoints:
+        - targetPort: 9402
+          path: "/metrics"
+          interval: 10s
+          scrapeTimeout: 5s
+      ```
+
+   5. ServiceAccount required by the agent, which will be associated with the required ClusterRole and
+      Roles in next step.
+      ```yaml
+      apiVersion: v1
+      kind: ServiceAccount
+      metadata:
+        labels:
+          app: cert-manager-istio-csr
+        name: cert-manager-istio-csr
+        namespace: istio-system
+      ```
+
+   6. ClusterRoles and Roles required by istio-csr.
+      ```yaml
+      ## required for creating ConfigMap with the CA in selected namespaces
+      ## by istio-csr agent.
+      kind: ClusterRole
+      apiVersion: rbac.authorization.k8s.io/v1
+      metadata:
+        name: cert-manager-istio-csr
```

Comment:
We may need a suffix for cluster role names to support multi-tenancy in the future.